### PR TITLE
[factorization] Update documentation, review examples + extras

### LINKED ISSUE #112: [factorization] m4 macro for CUDA
[These lines of code](https://github.com/barbagroup/PetIBM/blob/factorization/m4/configure_cuda.m4#L33-L37) in the m4 macro for CUDA will use `$CUDA_DIR/lib` in priority, if `$CUDA_DIR/lib` exists. However, on systems like Colonial One, `$CUDA_DIR/lib` and `$CUDA_DIR/lib64` both exist, while only `$CUDA_DIR/lib64` works. Making `$CUDA_DIR/lib` the first choice will stop configuration on systems like Colonial One.

### PR DESCRIPTION
**Pull request by mesnardo**
* Re-organize `m4` macros and fix #112.
* House keeping of the Makefile.am to standardize them.
* Fix warnings at compilation stage.
* In application 'navierstokes', fix bug when looping over the diffusive and convective terms when assembling the RHS for the velocity system (executable used to return an illegal memory access when using an implicit Euler scheme for the diffusive terms).
* Update examples to fit into the factorization branch.
* Add 2D example of the inline-oscillating cylinder (Re=100) with the decoupled ibpm.
* Update users documentation README to adapt to factorization branch.
* Remove old and unused post-processing Python scripts.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,31 +2,56 @@
...
-Currently, the system can be solved on CPUs using PETSc or on multiple GPUs using the Nvidia [AmgX](https://developer.nvidia.com/amgx) library and [AmgXWrapper](https://github.com/barbagroup/AmgXWrapper).
+Currently, two immersed boundary methods are implemented:
 
+* Immersed Boundary Projection Method (IBPM; Taira and Colonius, 2007);
+* decoupled version of the IBPM (Li et al., 2016).
 
-PetIBM solves the incompressible Navier-Stokes equations in two and three dimensions using the immersed boundary projection method from [Taira and Colonius](http://colonius.caltech.edu/pdfs/TairaColonius2007.pdf) (2007) and is implemented using [PETSc](http://www.mcs.anl.gov/petsc/), the Portable, Extensible Toolkit for Scientific Computation.
+PetIBM runs on distributed-memory architectures and relies on the [PETSc](http://www.mcs.anl.gov/petsc/) library for data structures and parallel routines.
+
+PetIBM solves the different linear systems either on CPUs using PETSc KSP objects or on multiple CUDA-capable GPU devices using the Nvidia [AmgX](https://github.com/NVIDIA/AMGX) library.
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply README.md]: Done in e05110e.

File: README.md [hunk 2]
```diff
@@ -2,31 +2,56 @@
...
+The data transfer between PETSc and AmgX is handled by the wrapper AmgXWrapper](https://github.com/barbagroup/AmgXWrapper).
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply README.md]: Done in e05110e.

File: README.md [hunk 3]
```diff
@@ -2,31 +2,56 @@
...
+
+PetIBM runs only on Unix-based systems (no support on Windows) and was last tested on Ubuntu 16.04 and MacOS Sierra 10.12.6.
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply README.md]: Yes, you should once the branch is merged.

File: README.md [hunk 4]
```diff
@@ -2,31 +2,56 @@
...
+* Matplotlib (2.0.2)
+
+Note: Python and libraries have been installed using Anaconda (4.4.0).
 
-Note: Python and libraries have been installed using [`conda`](http://conda.pydata.org/docs/get-started.html) (4.3.1).
+---
+
+## Documentation
+
+User's documentation is available on the [Wiki](https://github.com/barbagroup/PetIBM/wiki) pages of the cuIBM repository.
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply README.md]: Done in e05110e.

File: doc/inputs.md [hunk 5]
```diff
@@ -1,16 +1,236 @@
...
+
+This node gather various `parameters` regarding the advancement of a simulation.
+
+- `dt`: the time-step size.
+- `startStep`: index of the stating time step (default is zero).
+- `nt`: number of time steps to compute.
+- `nsave`: frequency (in number of time steps) of saving for the numerical solution.
+- `nrestart`: frequency (in number of time steps) of saving for the convective and diffusive terms; those terms will required upon restart of a run at a time step different from 0.
+- `convection`: time scheme for the convective terms; choices are the default explicit Euler method (`EULER_EXPLICIT`) or an explicit second-order Adams-Bashforth scheme (`ADAMS_BASHFORTH_2`).
+- `diffusion`: time scheme for the diffusive terms; choices are the default implicit Euler method (`EULER_IMPLICIT`), an explicit Euler method (`EULER_EXPLICIT`), or a second-order Crank-Nicolson scheme ('CRANK_NICOLSON').
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply doc/inputs.md]: Done in e05110e.

File: doc/installation.md [hunk 6]
```diff
@@ -1,17 +1,31 @@
...
 Ensure that the following dependencies are installed before compiling PetIBM:
 
-* GNU C++ Compiler(`g++`) version 4.6 or above
-* [PETSc](http://www.mcs.anl.gov/petsc/) version 3.7
+* GNU C++ compiler g++ (4.9.2, 5.4.0)
+* [PETSc](https://www.mcs.anl.gov/petsc/) (3.8.1)
+
+Optional to solve linear systems on GPUs:
+
+* NVIDIA's CUDA compiler nvcc (8.0)
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply doc/installation.md]: Done in e05110e.

File: doc/installation.md [hunk 7]
```diff
@@ -25,55 +39,73 @@ On Mac OS X, `g++` can be installed on Mac OS X via XCode or [Homebrew](brew.sh)
...
+        --with-cxx=g++ \
+        --with-fc=gfortran \
+        --COPTFLAGS="-O0" \
+        --CXXOPTFLAGS="-O0" \
+        --FOPTFLAGS="-O0" \
         --with-debugging=1 \
         --download-fblaslapack \
-        --download-mpich
+        --download-openmpi \
+        --download-hypre \
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply doc/installation.md]: Remove `--download-hypre`. Done in e05110e.

File: doc/installation.md [hunk 8]
```diff
@@ -101,58 +155,93 @@ and restart you terminal or  source the file:
...
+    PETSC_DIR=$HOME/sfw/petsc/3.8.1
+    PETSC_ARCH=linux-opt
     mkdir petibm-linux-opt && cd petibm-linux-opt
     $PETIBM_DIR/configure \
         --prefix=$HOME/sfw/petibm/petibm-linux-opt \
         CXX=$PETSC_DIR/$PETSC_ARCH/bin/mpicxx \
-        CXXFLAGS="-g -O3 -std=c++11" \
-        --with-amgx="$HOME/sfw/amgx" \
-        --with-cuda="/usr/local/cuda-6.5"
+        CXXFLAGS="-O3 -Wall -Wno-deprecated -std=c++14" \
```
[the remark this answers is not in the document]
Comment (mesnardo) [reply doc/installation.md]: No need for this one.

**Comment by piyueh** [comment]
It seems good to go.

**Pull request merged**